controllers/storageclaim_controller.go

```diff
@@ -690,3 +693,92 @@ func (r *StorageClaimReconciler) migrateStorageClassClaims() error {
 
 	return nil
 }
+
+func (r *StorageClaimReconciler) hasStorageClaimDependents() (bool, error) {
+
+	pvList := &corev1.PersistentVolumeList{}
+	if err := r.list(pvList, client.MatchingFields{pvClusterIDIndexName: r.storageClaim.Name}, client.Limit(1)); err != nil {
+		return false, fmt.Errorf("failed to list persistent volumes: %v", err)
+	}
+
+	// shortest path for verifying dependents overall
+	if len(pvList.Items) != 0 {
+		r.log.Info(fmt.Sprintf("PersistentVolumes referring storageclaim %q exists", r.storageClaim.Name))
+		return true, nil
+	}
+
+	// unlike pv which has a direct mapping available from it's spec to storageclaim, volumesnapshots doesn't have such a mapping
+	// so we need to follow the intermediate mappings to verify any snapshot is dependent on storageclaim or not
+	vsContentList := &snapapi.VolumeSnapshotContentList{}
+	if err := r.list(vsContentList); err != nil {
+		return false, fmt.Errorf("failed to list volume snapshot content resources: %v", err)
+	}
+
+	for idx := range vsContentList.Items {
+
+		vsContent := vsContentList.Items[idx]
+		driver := vsContent.Spec.Driver
+		if driver != csi.GetRBDDriverName() || driver != csi.GetCephFSDriverName() {
+			// skip the resources that doesn't belong to ceph csi
+			continue
+		}
+
+		vsClass := &snapapi.VolumeSnapshotClass{}
+		// shortest path for verifying snapshot dependents via volumesnapshotclass
+		if vsContent.Spec.VolumeSnapshotClassName != nil {
+			vsClass.Name = *vsContent.Spec.VolumeSnapshotClassName
```

Comment:
This may not be correct, extract the clusterID from the snapshotHandle that will cover all the cases

Reply:
`vsc.spec.source.volumeHandle` or `vsc.status.snapshotHandle`?

`volumeHandle: 0001-001b-ocs-storagecluster-ceph-rbd-0000000000000004-8cc378f9-7620-4eef-aa47-45372edecba9` vs `snapshotHandle: 0001-001b-ocs-storagecluster-ceph-rbd-0000000000000004-2360c9a6-46fa-4d31-b0e4-79104c1e2e65`

and what's the criteria to extract the clusterID?

btw, thanks a lot, good thing I didn't went down the rabbit hole and implement all those branches 😅

Reply:
`vsc.status.snapshotHandle` is the right one,

https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/clusterid-mapping.md#design-to-handle-clusterid-and-poolid-for-dr will give you idea on how to extract clusterID

Reply:
ack, i checked above, is there any func that actually does the transformation rather than duplicating the code?

Reply:
done.